synapseclient/models/dataset.py

```diff
@@ -3,63 +3,579 @@
+
+            synapse_client: If not passed in and caching was not disabled by
+                `Synapse.allow_client_caching(False)` this will use the last created
+                instance from the Synapse class constructor
+
+            **kwargs: Additional arguments that are passed to the `pd.DataFrame`
+                function when the `values` argument is a path to a csv file.
+
+
+        Example: Update custom column values in a dataset.
 
+            ```python
+            from synapseclient import Synapse
+            from synapseclient.models import Dataset
+
+            syn = Synapse()
+            syn.login()
+
+            my_dataset = Dataset(id="syn1234").get()
+
+            # my_annotation must already exist in the dataset as a custom column
+            modified_data = pd.DataFrame(
+                {"id": ["syn1234"], "my_annotation": ["good data"]}
+            )
+            my_dataset.update_rows(values=modified_data, primary_keys=["id"], dry_run=False)
+            ```
+        """
+        return None
+
+    def snapshot(
+        self,
+        *,
+        comment: Optional[str] = None,
+        label: Optional[str] = None,
+        activity: Optional[Activity] = None,
+        synapse_client: Optional[Synapse] = None,
+    ) -> "TableUpdateTransaction":
+        """Creates a snapshot of the dataset. A snapshot is a saved, read-only version of the dataset
+        at the time it was created. Dataset snapshots are created using the asyncronous job API.
+
+        Arguments:
+            comment: A unique comment to associate with the snapshot.
+            label: A unique label to associate with the snapshot.
+            activity: The Activity model represents the main record of Provenance in
+                Synapse. It is analogous to the Activity defined in the
+                [W3C Specification](https://www.w3.org/TR/prov-n/) on Provenance.
+            synapse_client: If not passed in and caching was not disabled by
+                `Synapse.allow_client_caching(False)` this will use the last created
+                instance from the Synapse class constructor.
 
-@dataclass()
+        Returns:
+            A `TableUpdateTransaction` object which includes the version number of the snapshot.
+
+        Example: Save a snapshot of a dataset.
+
+            ```python
+            from synapseclient import Synapse
+            from synapseclient.models import Dataset
+
+            syn = Synapse()
+            syn.login()
+
+            my_dataset = Dataset(id="syn1234").get()
+            my_dataset.snapshot(comment="My first snapshot", label="My first snapshot")
+            ```
+        """
+        return TableUpdateTransaction
+
+    @staticmethod
+    def query(
+        query: str,
+        include_row_id_and_row_version: bool = True,
+        convert_to_datetime: bool = False,
+        *,
+        synapse_client: Optional[Synapse] = None,
+        **kwargs,
+    ) -> DATA_FRAME_TYPE:
```

Comment:
I don't think that this is needed since I am going to be keeping `QueryMixinSynchronousProtocol` logic added here:

https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1181/files#r1991825702